### ISSUE TITLE
Timeout expired before operation completed

### ISSUE THREAD

**Issue by zfthink** [pfroud/JVisa#8]
Hello,
I tried to use JVisa to connect to a device, when I use

```java
jVisaInstrument.sendAndReceiveString("*IDN?\\n");
```

something is wrong, the trace is:

```
/project_code/jj/jvisa/target/classes:/root/.m2/repository/net/java/dev/jna/jna/5.11.0/jna-5.11.0.jar:/data/app/idea-IU-211.6693.111/lib/idea_rt.jar xyz.froud.jvisa_example.lowlevel.FindResourcesExample
Connected to the target VM, address: '127.0.0.1:38753', transport: 'socket'
Exception in thread "main" xyz.froud.jvisa.JVisaException: in viRead(): Timeout expired before operation completed. (0xBFFF0015)
	at xyz.froud.jvisa.JVisaResourceManager.checkError(JVisaResourceManager.java:291)
	at xyz.froud.jvisa.JVisaInstrument.readBytes(JVisaInstrument.java:152)
	at xyz.froud.jvisa.JVisaInstrument.readString(JVisaInstrument.java:168)
	at xyz.froud.jvisa.JVisaInstrument.sendAndReceiveString(JVisaInstrument.java:74)
	at xyz.froud.jvisa_example.lowlevel.FindResourcesExample.main(FindResourcesExample.java:37)
Disconnected from the target VM, address: '127.0.0.1:38753', transport: 'socket'
```

**Comment by zfthink** [comment]
i say pyvisa tell that set  
**power.write_termination = "\n"
power.read_termination = "\n"**
where can i set the same thing in jvisa

**Comment by pfroud** [comment]
Hi, I can help you debug this issue. 

Can you tell me what OS and what VISA implementation you are using?

Right now I do not know about the PyVISA `write_termination` and `read_termination` settings. I will read the PyVISA code to learn about it

**Comment by zfthink** [comment]
> Hi, I can help you debug this issue.
> 
> Can you tell me what OS and what VISA implementation you are using?
> 
> Right now I do not know about the PyVISA `write_termination` and `read_termination` settings. I will read the PyVISA code to learn about it

thank you a lot.
the OS is  centos7.X ,and NI-VISA

**Comment by zfthink** [comment]
> Hi, I can help you debug this issue.
> 
> Can you tell me what OS and what VISA implementation you are using?
> 
> Right now I do not know about the PyVISA `write_termination` and `read_termination` settings. I will read the PyVISA code to learn about it

by the way ,here is a link of pyvisa issue *https://github.com/pyvisa/pyvisa/issues/397*

**Comment by pfroud** [comment]
First I want to check if communication with your device works using other libraries. 

* Does it work using PyVisa?
* Can you try using [VISA Interactive Control](https://www.ni.com/docs/en-US/bundle/ni-visa/page/ni-visa/interactivecontrolvisa.html):
    * On Linux the executable is called `NIvisaic`. 
    * After you launch it, first select the instrument:
      <img src="https://user-images.githubusercontent.com/9056906/218627221-f2adcda9-5e50-45a4-9a2e-d70de37cb289.png" alt="" width="50%">

    * On the top, click "Input/Output":
    ![VISA IC 02](https://user-images.githubusercontent.com/9056906/218627534-24606941-2d6b-416e-b16b-ac5c7aabd3ee.png)
    * The text field will be populated with "*IDN?\n", click "Query":
    ![screenshot of VISA Interactive Control](https://user-images.githubusercontent.com/9056906/218627983-928642c2-2f7d-42b0-bf65-d123d63ef833.png)

**Comment by zfthink** [comment]
> First I want to check if communication with your device works using other libraries.
> 
> * Does it work using PyVisa?
> * Can you try using [VISA Interactive Control](https://www.ni.com/docs/en-US/bundle/ni-visa/page/ni-visa/interactivecontrolvisa.html):
>   
>   * On Linux the executable is called `NIvisaic`.
>   * After you launch it, first select the instrument:
>     <img alt="" width="50%" src="https://user-images.githubusercontent.com/9056906/218627221-f2adcda9-5e50-45a4-9a2e-d70de37cb289.png">
>   * On the top, click "Input/Output":
>     ![VISA IC 02](https://user-images.githubusercontent.com/9056906/218627534-24606941-2d6b-416e-b16b-ac5c7aabd3ee.png)
>   * The text field will be populated with "*IDN?\n", click "Query":
>     ![screenshot of VISA Interactive Control](https://user-images.githubusercontent.com/9056906/218627983-928642c2-2f7d-42b0-bf65-d123d63ef833.png)

because of something special ，i can only use JAVA  not python.
i have already use the tools of ni-visa in linux ,and it works correct

**Comment by zfthink** [comment]
> > First I want to check if communication with your device works using other libraries.
> > 
> > * Does it work using PyVisa?
> > * Can you try using [VISA Interactive Control](https://www.ni.com/docs/en-US/bundle/ni-visa/page/ni-visa/interactivecontrolvisa.html):
> >   
> >   * On Linux the executable is called `NIvisaic`.
> >   * After you launch it, first select the instrument:
> >     <img alt="" width="50%" src="https://user-images.githubusercontent.com/9056906/218627221-f2adcda9-5e50-45a4-9a2e-d70de37cb289.png">
> >   * On the top, click "Input/Output":
> >     ![VISA IC 02](https://user-images.githubusercontent.com/9056906/218627534-24606941-2d6b-416e-b16b-ac5c7aabd3ee.png)
> >   * The text field will be populated with "*IDN?\n", click "Query":
> >     ![screenshot of VISA Interactive Control](https://user-images.githubusercontent.com/9056906/218627983-928642c2-2f7d-42b0-bf65-d123d63ef833.png)
> 
> because of something special ，i can only use JAVA not python. i have already use the tools of ni-visa in linux ,and it works correct

**Comment by pfroud** [comment]
OK, I will try to implement `write_termination` and `read_termination` in JVisa

**Comment by zfthink** [comment]
oh ,thank you ,please tell me if you have finished it .

**Comment by pfroud** [comment]
Please try removing the second backslash in your IDN query:
```java
/* before */ jVisaInstrument.sendAndReceiveString("*IDN?\\n");
/* after  */ jVisaInstrument.sendAndReceiveString("*IDN?\n");
```

**Comment by zfthink** [comment]
i have already try it before i send this issue ,unfortunately，it did not work !

**Comment by pfroud** [comment]
I created branch [add-termination-chars](https://github.com/pfroud/JVisa/tree/add-termination-chars). I have not tested it, but please try this code:

```java
jVisaInstrument.setReadTerminationCharacter('\n');
jVisaInstrument.setReadTerminationCharacterEnabled(true);
jVisaInstrument.setWriteTerminator("\n");
jVisaInstrument.sendAndReceiveString("*IDN?");
```

Let me know if you need help using Git branches

**Comment by zfthink** [comment]
thank you again , i will test it  and tell you weather it work later

**Comment by zfthink** [comment]
hello,my brother, Sorry to have kept you waiting.

thank you again.it worked! After adding terminationCharacter settings ,i can get result from device.

But,here have another problem , how can i find resources from another ip (not localhost)?

can you add a constructor  like **resourceManager.findResources("xxx.xxx.xxx.xxx");** to implement it ?

Looking forward to your reply

**Comment by pfroud** [comment]
Good news, thank you for testing it!

Do you mean you want to find resources connected to another computer (not
localhost)?

**Comment by zfthink** [comment]
yeah，I want to control or get result from  device  by Ethernet .

At present ,I use **resourceManager.openInstrument("TCPIP0::192.168.1.11::1799::SOCKET");** to implement . In fact , I want to get resources from another IP first .

**Comment by zfthink** [comment]
oh , when I test it ,here still have something wrong :

**Comment by pfroud** [comment]
I don't know about opening multiple sessions to a device at the same time.

One of the arguments to [`viOpen()`](https://www.ni.com/docs/en-US/bundle/ni-visa/page/ni-visa/viopen.html) is `accessMode`. I don't know what access mode the NI tools use to open instruments. JVisa uses null (zero) which means:

> the session uses VISA-supplied default values.

The page for [`viLock()`](https://www.ni.com/docs/en-US/bundle/ni-visa/page/ni-visa/vilock.html) it says:

> If a session locks a resource, operations invoked by other sessions to the same resource are serviced or returned with a locking error, depending on the operation and the type of lock used.

I think a locking error is different from the "insufficient location information or not present" error you got.


Please open a separate issue for your question about finding resources from another IP.

**Comment by pfroud** [comment]
Support for termination characters merged into the main branch in commit ae2fc0db1e8aee54c650d067406c6bc29235f5ec
